### PR TITLE
refactor(memory): GraphStore seam + visibility parity for graph traversal

### PR DESCRIPTION
**Pull request by WingedGuardian**
PR-1 of two for the graph-engine migration (issue #1641). Two commits, deliberately separate: a **seam** that changes no behaviour, and a **behaviour change** that can be reviewed and reverted on its own.

## Commit 1 — the seam

The memory graph is an in-process NetworkX projection of `memory_links`. Swapping it for a server-backed engine meant touching every reader, so this adds the seam that engine plugs into: a `GraphStore` protocol, the incumbent moved behind it as `NetworkxGraphStore`, and `graph.py` reduced to a facade owning the backend choice and the single production instance.

**The contract is the point.** A read that cannot *reach* its store raises `GraphUnavailableError` and never returns empty — empty must keep meaning empty. Collapsing the two is how a missing NetworkX import once disarmed the importance shield: its empty result read as "no bridge memories", wiping `centrality_cache`. Traversal still degrades to the recursive CTE and now says so in the log; centrality still refuses to degrade at all. CI runs no type checker, so a conformance **test** asserts the contract rather than an annotation alone — PR-2's backend inherits it.

**One behaviour change rides here, and it closes a real gap:** the cached projection now notices writes from *other* processes. It only ever tracked a flag flipped in the writing process, so the server could serve a graph predating every dream-job write until it happened to write a link itself. It now also compares SQLite's `data_version` — stamped **before** the load, deliberately: in WAL the read snapshot is fixed at the first SELECT step, so stamping afterwards pins a projection missing a mid-load commit. That was measured, and it loses the write permanently.

Behaviour-preserving otherwise, measured not assumed: **400 real roots at the real call parameters, old path vs new, 0/400 divergence** in both node sets and top-five slices.

## Commit 2 — visibility parity (a deliberate live behaviour change)

`search_ranked` hides bitemporally-expired and deprecated memories; `graph_expansion` repeats that filter citing "visibility parity with normal recall". **`traverse` did neither**, and `mcp/memory/core.py` emits raw memory_ids into `graph_neighbors` with no hydration — so memories recall deliberately suppresses were reaching the model as live context.

The predicate now lives once in the seam and every backend applies it, CTE fallback included, **root as well as neighbours**. It ignores `valid_at` on purpose: the bake-off oracle reads it but answers an as-of question, and copying it would hide the 5.7% of rows with a NULL `valid_at`.

Measured on the live graph:

| | |
|---|---|
| edges dropped | 29,868 / 264,191 (11.3%) |
| top-5 slices changed | 107/400 (26.8%) |
| reach-sets changed | 219/400 (54.8%) |
| roots whose neighbours were *entirely* hidden memories | 26/400 (6.5%) |
| roots that GREW | **0/400** |

That last row is the direction invariant; the 6.5% is enrichment that was pure noise.

**Second blast radius, declared:** filtering the source endpoint also shrinks the graph centrality runs on, so the shield's bridge-node population moves — measured **15,771 → 12,670** nonzero scores, far clear of the empty case that would *silently* disable bridge protection. That covers the whole graph, not a single dream slice; MW-5 inherits the difference.

Implemented as two queries plus set membership rather than one SQL pass: identical sets, but **+2.8s per cold build vs +99ms**, and the new staleness token makes rebuilds more frequent.

## Verification

- **123 tests** green across 8 files; two independent verify-RED passes (removing the filter fails 3 tests; removing the CTE root clause fails the hidden-root parity test, which carries a live-root control).
- Two adversarial audits. The first found a **blocker** in the token's snapshot ordering — reproduced independently before acting, with a refutation control. The second found the **CTE/NX divergence at the root** (2,827 live memories are hidden *and* have out-edges; 6,503 edges classified differently), a now-false comment in `dream_shield`, and a predicate narrower than the readers it mirrors (`deprecated = 1` → `!= 0`).
- Docs updated: `CURRENT.md` records the seam, the contract, the measured blast radius, and the freshness model's real boundary — time-driven expiry arrives with no write event, so a quiet process can serve an expired memory until an unrelated rebuild (exposure: 114 memories, 4 with edges). CHANGELOG carries a user-facing entry.

## Not in this PR

`SqliteGraphStore`, the config module and fallback wiring are PR-2. No FalkorDB install, no projector, no migration of the readers that bypass `graph.py`.

---

### Ordering note vs #1628 (added 2026-09-06)

This PR moves `_bfs_with_strength` into `graphstore_nx.py` **verbatim**, which
means it carries the pre-fix per-parent BFS. #1628 fixes that same function in
`graph.py` (level-synchronous best-parent-wins: ~6% of reported nodes gain a
truer strength, none are ever lowered, and the top-five slice stops depending
on loader row order). The two therefore conflict as modify/delete.

**Either merge order is safe** — #1628's tests drive the public `traverse()`
facade this PR keeps, so a resolution dropping that fix fails loudly.

Preferred order is **#1628 first**, then rebase this branch so the move carries
the already-fixed function. That keeps the "pure behaviour-preserving move"
claim above literally true; re-run the 400-root replay after the rebase, since
its baseline shifts with it. If this PR lands first, #1628 gets re-targeted to
`graphstore_nx.py` instead.

## Summary by CodeRabbit

* **New Features**
  * Added graph-based memory traversal and centrality analysis.
  * Added NetworkX-backed processing with typed parallel links and strength filtering.
  * Added cache reuse and cross-connection update detection for graph data.
* **Bug Fixes**
  * Graph results now consistently exclude deprecated, expired, or otherwise invalid memories.
  * Improved traversal fallback when graph processing is unavailable.
* **Tests**
  * Expanded coverage for visibility filtering, cache invalidation, fallback traversal, centrality availability, and parallel links.

### CODE REVIEW

**Comment by WingedGuardian** [comment]
@codex review

**Comment by WingedGuardian** [comment]
**Automated first-pass review — structural + PII/secrets/wording**

Reviewed head `9e84fa1d53a946e4b9ff80fdb887c3de74069c72` (5 commits, 15 files, +1416/-260). Not doc-only, so both passes ran.

### Structural

No pattern violations found: no `nohup`, no direct `task_states` inserts, no `/tmp/` usage (tests correctly use pytest's `tmp_path`), no `git add .`/`-A`, no force-push/reset patterns, no bare `except:` (all typed `except Exception:`/`except ImportError:`), no mutable default arguments, no `os.killpg` misuse. Import hygiene is clean throughout (`graph.py`, `graphstore.py`, `graphstore_nx.py`): stdlib → local grouping is correct, no star imports, no circular imports (`graphstore.py` has zero imports from its two new consumers). All 4 substantive commits carry conventional prefixes (`fix:`/`refactor:`).

Two lower-severity items:

- **Warning — raw SQL against `genesis.db` outside the CRUD layer, in three places now instead of two.** `graphstore.py`'s new `_INVALID_MEMORY_SQL`/`invalid_memory_ids()` (graphstore.py:~65-92), the `SELECT ... FROM memory_links` bulk loader in `graphstore_nx.py` (~line 1018 of the loader), and the new `NOT EXISTS` subqueries added to `_traverse_cte` in `graph.py` all query `memory_metadata`/`memory_links` directly rather than through `genesis.db.crud`. This matches the pre-existing pattern in this file (bulk/recursive graph reads aren't CRUD-shaped), so not requesting a change — but the PR body itself notes the same visibility predicate is now expressed independently in `db/crud/memory.py::search_ranked`, `memory/graph_expansion.py`, and this new raw SQL. Three independent expressions of one predicate is a real drift risk (a future edit to one that misses the others reopens exactly the parity gap this PR closes) — worth a follow-up to consolidate, even if not a blocker here.
- **Warning — new file `src/genesis/memory/graphstore.py` has no dedicated test file.** It's exercised indirectly (`test_networkx_store_satisfies_the_seam` in `tests/test_memory/test_graphstore_nx.py`, fixtures in `test_graph.py`), but `invalid_memory_ids()` and the `GraphUnavailableError`/`GraphStore` definitions have no `tests/test_memory/test_graphstore.py` of their own. Likely fine given the coverage is real, just flagging the missing 1:1 file mapping.

Docstrings are present on all new public surface (`GraphStore`, `GraphNode`, `TraversalResult`, `NetworkxGraphStore` and its methods, `invalid_memory_ids`), and the `dream_shield.py` comment update accurately reflects the new upstream filtering.

### PII/Secrets/Wording

Scanned all added lines plus PR title/body and the `CHANGELOG.md` addition. No API keys/tokens/secrets, no private IPv4/IPv6 ranges, no SSH keys, no connection strings, no personal emails, no user-bearing absolute paths, no location-revealing timezones, no cloud account IDs, no hardware identifiers, no private hostnames. PR/changelog wording describes an internal memory-visibility correctness fix (expired/deprecated memories reaching recall) — not phrased as a PII/secrets leak, security incident, or exposed-infrastructure disclosure, and names no private infrastructure.

PII/Secrets/Wording: CLEAN

VERDICT: PASS



---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by WingedGuardian** [comment]
@codex review

File: src/genesis/memory/graphstore_nx.py [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+        )
+        if G.number_of_edges() > 50_000:
+            logger.warning(
+                "Graph has %d edges — measure NetworkX rebuild cost; consider an "
+                "incremental or server-backed graph if rebuilds become a bottleneck",
+                G.number_of_edges(),
+            )
+
+        self._graph = G
+        self._dirty = False
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/memory/graphstore_nx.py:424]: Verified this rather than applying the patch, and the mechanism does not hold.

The premise is right: `PRAGMA data_version` does not move for a commit on our own connection. Measured on aiosqlite + WAL — unchanged after this connection commits, incremented after another one does. The reachability is right too: Genesis shares one connection across subsystems and `SerializedConnection`'s lock is per-method, so a writer really can commit inside the load window.

The step that fails is the assumption that the same-connection write is *absent* from the load. Measured in that exact window:

- commit on the **same** connection between `execute()` and `fetchall()` → `rows=['A','B']` — **visible** to that fetch
- identical commit from **another** connection → **invisible**, snapshot held

A connection reads its own writes, so the racing same-connection commit is already in the rows the graph is built from. Clearing `_dirty` discards nothing, because there is no staleness left to signal. The two signals cover complementary cases with no gap: the token catches the other-connection commit (already locked by `test_a_commit_inside_the_load_window_is_not_lost`), and the same-connection one needs no catching.

There is a real residual, narrower than reported. Today's correctness depends on there being **no `await` between `await cursor.fetchall()` and `self._dirty = False`** — everything in between is synchronous, so nothing can interleave. Add an await there and this becomes a genuine bug, for the same-connection case only. That invariant was unstated, which is the actual defect here.

Addressed in 251033e7c without a behaviour change: a characterization test pinning the measured visibility (verified non-vacuous by mutation — swapping the injection to another connection makes it fail), and the invariant stated at the line where breaking it would matter.

**Comment by WingedGuardian** [comment]
@codex review

**Comment by WingedGuardian** [comment]
@codex review

**Comment by WingedGuardian** [comment]
@codex review

**Comment by WingedGuardian** [comment]
@codex review

File: src/genesis/memory/graphstore_nx.py [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+        # connection per spawn. A second long-lived connection appearing in any
+        # of those processes would turn this into a rebuild storm on a
+        # 264k-edge graph (seconds per rebuild) — that invariant is
+        # load-bearing and worth re-checking before adding one.
+        if db is not self._built_conn:
+            return True
+        if self._built_data_version is None:
+            return False
+        current = await self._data_version(db)
+        return current is not None and current != self._built_data_version
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/memory/graphstore_nx.py:315]: Confirmed, and accepted deliberately rather than patched here.

`PRAGMA data_version` moves when **any** other connection commits, so an unrelated write to the shared file does force a rebuild of a 264k-edge graph. The finding is right about the mechanism.

Two reasons it stays:

The error is in the safe direction. Over-rebuilding costs time; under-rebuilding serves a stale graph, which is the failure this token exists to prevent. A cheaper signal that can miss a `memory_links` write would be worse than the cost being described.

The real fix is a DB-side generation token scoped to the table, which is filed as #1641 and is the next slice of this program. It serves two consumers — this staleness check and the projector's dirty signal — so it gets built once, deliberately, rather than twice by accident. A narrow fix here would be replaced within days and would mean touching a PR that is otherwise ready.

Answered-by-plan rather than ignored: the cost is real, the direction is correct, and the durable fix is scheduled and scoped.

File: src/genesis/memory/graphstore_nx.py [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+        #    incidentally (its writers rewire links); TIME-DRIVEN expiry is not
+        #    — an `invalid_at` in the future arrives with no write event at all,
+        #    so a quiet process keeps serving the memory until some unrelated
+        #    rebuild. MEASURED exposure: 114 memories carry a future invalid_at,
+        #    4 of them have any edge.
+        #  - These are two autocommit statements, hence two WAL read snapshots;
+        #    an edge committed between them is filtered against an invalid-set
+        #    read just before it. Self-healing — pre_load_version is stamped
+        #    ahead of BOTH, so the next read rebuilds.
+        invalid = await invalid_memory_ids(db)
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/memory/graphstore_nx.py]: Real, and already documented in the file with its measured exposure — this is a stated caveat rather than an oversight.

`graphstore_nx.py` says it directly: every `invalidate_graph_cache()` site is a `memory_links` writer, while this predicate reads `memory_metadata`. Deprecation is covered incidentally, because its writers rewire links; a metadata-only change is not.

Accepting it rather than widening the invalidation surface, because the alternative is worse: invalidating on every metadata write would rebuild the graph for changes that cannot affect it, and that cost lands on the recall path. The durable fix is the same table-scoped generation token as the P1 above (#1641), which is why both are answered together rather than separately.

File: src/genesis/memory/graphstore_nx.py [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+        # passes memory_mod._db, dream-centrality passes the runtime's shared
+        # serialized connection, and the ambient worker opens one mode=ro
+        # connection per spawn. A second long-lived connection appearing in any
+        # of those processes would turn this into a rebuild storm on a
+        # 264k-edge graph (seconds per rebuild) — that invariant is
+        # load-bearing and worth re-checking before adding one.
+        if db is not self._built_conn:
+            return True
+        if self._built_data_version is None:
+            return False
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/memory/graphstore_nx.py]: Confirmed by reading the source, and **fixed** in `6be3ae3b5`.

`SerializedConnection`'s recovery path closes the wrapped handle and rebinds `_conn` via `object.__setattr__`, so the proxy's own identity never changes. The check could therefore never fire, and the `data_version` comparison behind it then read a fresh connection's counter against one stamped from the connection that had just been closed — unrelated per-connection values.

The consequence is worse than "no rebuild": staleness detection stops working at exactly the moment the database has just recovered from errors, which is when the projection is most likely to be out of date.

Fixed by tracking what the proxy wraps rather than the proxy. Unwrapping is uniform rather than a special case — a bare `aiosqlite.Connection` exposes the same attribute for its own handle, and both change precisely when the real connection underneath does. `_conn` is in `SerializedConnection._OWN_ATTRS`, so the read is never delegated onward.

Two tests: one drives a reconnect behind a stable proxy, one is the negative control that a check rebuilding on *every* call would fail. Mutation-verified against the genuine pre-fix state — reverting **both** sites turns the reconnect test red. Worth noting that reverting only one site does *not*, because a one-sided revert creates a permanent mismatch that rebuilds always; a mutation is only meaningful if it reproduces a state the code could actually be in.

File: src/genesis/memory/graphstore_nx.py [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+        #    read just before it. Self-healing — pre_load_version is stamped
+        #    ahead of BOTH, so the next read rebuilds.
+        invalid = await invalid_memory_ids(db)
+        cursor = await db.execute(
+            "SELECT source_id, target_id, link_type, strength FROM memory_links"
+        )
+        rows = [
+            row
+            for row in await cursor.fetchall()
+            if row[0] not in invalid and row[1] not in invalid
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/memory/graphstore_nx.py]: Confirmed against source, accepted here, and **filed as #1896** so it is tracked rather than dropped.

Verified before filing: `include_deprecated` is a documented public parameter threaded into the search, the traversal is called with no visibility override, and the commit that added the filtering is the one that introduced the interaction. The audit path did work before it — by accident, since `traverse` hid nothing at all.

Not fixed in this PR because honouring the flag means the visibility decision has to travel with the request instead of being fixed by the store: a signature change on the seam protocol and every implementation of it, plus the call sites. That is not a change to make on a PR that is one gate from merging, and the path is opt-in and rarely taken.

The issue carries the sketch, including the two things worth pinning with tests — that the default path stays byte-identical, and that a deprecated *neighbour* of a visible root is still hidden unless asked for, which is a different question from a deprecated *root* and easy to conflate.

File: src/genesis/memory/graphstore_nx.py [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+        #     races it (a connection reads its own writes), so the graph already
+        #     contains it and there is no staleness to signal. data_version
+        #     deliberately does not move for it, and does not need to.
+        # An await here would break the second case and only the second case: a
+        # same-connection writer could then commit AFTER the fetch, set _dirty,
+        # and have it cleared on the next line with nothing else left to notice.
+        # Locked by test_a_same_connection_write_is_seen_by_the_load_that_races_it.
+        self._dirty = False
+        self._built_conn = db
+        self._built_data_version = pre_load_version
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/memory/graphstore_nx.py:426]: Real, and one of the two freshness caveats this file states in-code with its measured exposure: 114 memories carry a future `invalid_at`, 4 of which have any edge.

The mechanism is that time-driven expiry arrives with no write event at all, so nothing invalidates and a quiet process keeps serving the memory until some unrelated rebuild. That is accurately described where the predicate is loaded.

Accepted rather than fixed, because the fixes available are both worse than the exposure: a TTL on the projection rebuilds a 264k-edge graph on a timer regardless of need, and a per-traversal re-check moves the visibility query onto the recall path it was deliberately taken off. Documented with a number rather than hidden.

File: tests/test_memory/test_graph_store_reset_wiring.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+    Identity, not state: a reset that rebuilt the same object in place would
+    still leave the old connection reachable, which is the thing being
+    prevented. These two tests must stay in this order and in this file — they
+    are one assertion split across the fixture boundary, because that boundary
+    is what is under test.
+    """
+    _seen.append(id(graph_mod._store))
+    assert len(_seen) == 2, "both halves must run; this test cannot stand alone"
+    assert _seen[0] != _seen[1], (
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply tests/test_memory/test_graph_store_reset_wiring.py]: Right, and **fixed** in `6be3ae3b5`.

Once the fixture replaces the module global, nothing refers to the first store, so CPython may hand its address to the one allocated immediately afterwards. Two distinct objects then compare equal and the test fails for a reason unrelated to the wiring it guards — and the same mechanism can mask a real regression by coincidence. It is unsound in both directions, which is what makes it worth fixing rather than tolerating.

It holds the objects now and compares with `is not`. The retention is two objects for the lifetime of one test module, which is not the leak this file exists to catch — that leak is a production global pinning a closed connection across a whole session, and it is unaffected by a test module briefly keeping its own reference. The comment says so, so the next reader does not "fix" it back.

File: tests/test_memory/test_graph_store_reset_wiring.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+from __future__ import annotations
+
+from genesis.memory import graph as graph_mod
+
+_seen: list[int] = []
+
+
+def test_the_store_identity_is_recorded_first():
+    """Half one: record which store object this test saw."""
+    _seen.append(id(graph_mod._store))
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply tests/test_memory/test_graph_store_reset_wiring.py]: Same defect as the Codex P2 on this file, and **fixed** in `6be3ae3b5` — recording it here too so this thread is not left hanging.

The test now holds the store objects and compares with `is not` rather than recording `id()` integers, which CPython is free to reuse once the first store becomes unreachable.

**Comment by WingedGuardian** [comment]
@codex review

Round 4, requested with the owner's explicit approval under the escalation cap. Round ledger and what changed:

| round | commit | outcome |
|---|---|---|
| 1 | `76e326e45bc7` | findings |
| 2 | `4b63f61cbb41` | findings |
| 3 | `987a918aa89f` | findings |

This round is not another pass over the same code. Since round 3 the branch has taken a merge of `origin/main` and two fixes that round 3 itself asked for, and the merge is the reason the review is worth spending:

**The reconcile is the diff nobody re-reviews.** #1628 changed `_bfs_with_strength` in place on main while this branch had moved that function into `graphstore_nx.py`. A textual merge can keep the move and discard the change with no conflict marker and a green suite. The carry was done by hand and proven by mutation — reverting to the per-parent walk turns four tests red — but that is exactly the kind of claim worth checking from outside.

**Two round-3 findings are fixed** (`6be3ae3b5`): the staleness check compared a `SerializedConnection` proxy whose identity is stable across a reconnect, so it could never fire; and a test compared `id()` integers that CPython may reuse.

**One finding an internal audit caught that no reviewer had:** the CTE's recursive step carries the visibility predicate, and nothing tested it — both existing parity tests build star graphs, so they only ever exercised the anchor. Neutering the recursive clause left all 49 tests green while the fallback returned a deprecated memory two hops out. A depth-two case now closes it.

**Dispositions posted on the four remaining threads**, all documented-accept with reasons rather than silence: the whole-database `data_version` counter (over-rebuilds, correct direction, real fix is #1641), two freshness caveats already stated in-code with measured exposure, and `include_deprecated`, which is filed as #1896 because honouring it means a signature change across the seam protocol.

The acceptance replay the PR body owes after a rebase was re-run against the live database: 400 highest-degree roots, NetworkX store against the CTE fallback, both filtered — identical node sets, identical top-five slices, identical full ordering.

Most useful place to look: the merge resolution in `src/genesis/memory/graph.py` and the hand-carried walk in `src/genesis/memory/graphstore_nx.py`.

**Comment by WingedGuardian** [comment]
@codex review

Re-issuing the round-4 request from earlier: it returned "Something went wrong. Try again later" rather than a review, so no round was consumed — the reviews API still shows three. Same approval, same head `6be3ae3b5ad0`, same scope as the previous comment: the merge resolution in `src/genesis/memory/graph.py` and the hand-carried walk in `src/genesis/memory/graphstore_nx.py` are where a fix could have silently vanished.

File: src/genesis/memory/graphstore_nx.py [hunk 9]
```diff
@@ -0,0 +1,455 @@
...
+    Unwrapping is uniform rather than a special case: a bare
+    `aiosqlite.Connection` also exposes `_conn` (its underlying sqlite3 handle),
+    and in both cases that attribute changes precisely when the real connection
+    underneath does. `_conn` is in `SerializedConnection._OWN_ATTRS`, so reading
+    it returns the proxy's own wrapped handle and is never delegated onward.
+
+    Two proxies sharing one underlying connection deliberately compare EQUAL —
+    same connection, same data, nothing to rebuild.
+    """
+    return getattr(db, "_conn", db)
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/memory/graphstore_nx.py]: Confirmed against the library rather than accepted on assertion, and **fixed** in `cd0386a5a`.

`aiosqlite.Connection._conn` is a property (`isinstance(type(c).__dict__["_conn"], property)` is `True`), not a plain attribute. Open, it returns the sqlite3 handle; closed, it raises `ValueError("no active connection")` — and `getattr(obj, name, default)` swallows only `AttributeError`, so the default never covered it.

Your reading of the severity is right and it is the part worth restating: because this runs in `_is_stale()`, upstream of the guarded reads, the raw error escaped `traverse()` and `graph.py` catches only `GraphUnavailableError` — so it bypassed the fallback chain entirely. That is the leak this PR's typed contract exists to close, reintroduced one layer above the guard that closed it.

Your diagnosis of why nothing caught it is also exactly right: the existing closed-connection test starts from a fresh store, so `_is_stale()` returns early on the empty cache and never reaches the probe.

The probe is now non-throwing by contract. Falling back to the object itself on a closed handle is correct rather than merely quiet — the token was captured while the connection was open, so the comparison fails, the projection reads stale, and the rebuild runs into the guarded reads and raises `GraphUnavailableError` as the contract requires. The failure is honoured through the guard rather than routed around it.

New test builds first so the cache is warm and the probe actually runs, then closes the handle and asserts the typed error from both `traverse()` and `centrality()`. Reverting the probe to the throwing form fails exactly that test and nothing else.

For the record on how it got in: I checked the attribute on an *open* connection, saw it present on both a bare connection and the proxy, and concluded the unwrap was uniform. I never closed one.
